src/cs/Ssh.Tcp/Services/PortForwardingService.cs

```diff
@@ -506,29 +537,35 @@ public PortForwardingService(SshSession session)
 			{
 				if (request.RequestType == PortForwardRequestType)
 				{
-					int? localPort;
 					try
 					{
-						localPort = await StartForwardingAsync(localEndPoint, cancellation)
+						localPort = await StartForwardingAsync(
+							localIPAddress,
+							(int)portForwardRequest.Port,
+							cancellation)
 							.ConfigureAwait(false);
 					}
 					catch (SocketException)
 					{
 						// The exception is already traced.
-						localPort = null;
 					}
 
 					if (localPort != null)
 					{
+						// The chosen local port may be different from the requested port. Use the
+						// requested port in the response, unless the request was for a random port.
```

Comment:
Why the requested port is not used if the request was for a random port?

Reply:
If the request was 0, indicating a random port, then there was no specific port number in the request. So the actual chosen port number is returned in the response and that number is used in further communications. In that case the resulting "remote" port number is the same as the "local" port number.